pkg/guacanalytics/patchPlanning.go

```diff
@@ -0,0 +1,179 @@
+		queue:   make([]string, 0), // the queue of nodes in bfs
+		nodeMap: map[string]DfsNode{},
+	}
+
+	q.nodeMap[startID] = DfsNode{
+		nodeType: startType,
+	}
+
+	q.queue = append(q.queue, startID)
+
+	for len(q.queue) > 0 {
+		q.now = q.queue[0]
+		q.queue = q.queue[1:]
+		q.nowNode = q.nodeMap[q.now]
+
+		if stopID == q.now {
+			break
+		}
+
+		if q.nowNode.depth >= maxDepth {
+			break
+		}
+
+		neighborsResponse, err := model.Neighbors(ctx, gqlclient, q.now, []model.Edge{})
+
+		if err != nil {
+			return nil, fmt.Errorf("failed getting neighbors:%w", err)
+		}
+
+		for _, neighbor := range neighborsResponse.Neighbors {
+			q, err = caseOnPredicates(ctx, gqlclient, &q, neighbor)
+
+			if err != nil {
+				return nil, err
+			}
+		}
+
+		q.nowNode.expanded = true
+		q.nodeMap[q.now] = q.nowNode
+	}
+
+	return q.nodeMap, nil
+
+}
+
+func caseOnPredicates(ctx context.Context, gqlclient graphql.Client, q *queueValues, neighbor model.NeighborsNeighborsNode) (queueValues, error) {
+	// case on predicates and nodeType
+	switch q.nowNode.nodeType {
+	case "packageVersion":
+		switch neighbor := neighbor.(type) {
+		case *model.NeighborsNeighborsIsDependency:
+			q, err := exploreIsDependency(ctx, gqlclient, q, *neighbor)
+
+			if err != nil {
+				return q, err
+			}
+		}
+	}
+
+	return *q, nil
+}
+
+func exploreIsDependency(ctx context.Context, gqlclient graphql.Client, q *queueValues, isDependency model.NeighborsNeighborsIsDependency) (queueValues, error) {
+	dependentPkgFilter := &model.PkgSpec{
+		Type:      &isDependency.DependentPackage.Type,
+		Namespace: &isDependency.DependentPackage.Namespaces[0].Namespace,
+		Name:      &isDependency.DependentPackage.Namespaces[0].Names[0].Name,
+	}
+
+	depPkgResponse, err := model.Packages(ctx, gqlclient, dependentPkgFilter)
+	if err != nil {
+		return *q, fmt.Errorf("error querying for dependent package: %w", err)
+	}
+
+	depPkgVersionsMap := map[string]string{}
+	var depPkgVersions []string
+	for _, depPkgVersion := range depPkgResponse.Packages[0].Namespaces[0].Names[0].Versions {
+		depPkgVersions = append(depPkgVersions, depPkgVersion.Version)
+		depPkgVersionsMap[depPkgVersion.Version] = depPkgVersion.Id
+	}
+
+	matchingDepPkgVersions, err := depversion.WhichVersionMatches(depPkgVersions, isDependency.VersionRange)
+	if err != nil {
+		return *q, fmt.Errorf("error determining dependent version matches: %w", err)
+	}
+
+	for matchingDepPkgVersion := range matchingDepPkgVersions {
+		matchingDepPkgVersionID := depPkgVersionsMap[matchingDepPkgVersion]
+		if err != nil {
+			return *q, fmt.Errorf("error querying neighbor: %w", err)
+		}
+
+		q.path = append(q.path, isDependency.Id, matchingDepPkgVersionID,
+			depPkgResponse.Packages[0].Namespaces[0].Names[0].Id, depPkgResponse.Packages[0].Namespaces[0].Id,
+			depPkgResponse.Packages[0].Id)
+
+		dfsNVersion, seenVersion := q.nodeMap[matchingDepPkgVersionID]
+		dfsNName, seenName := q.nodeMap[depPkgResponse.Packages[0].Namespaces[0].Names[0].Id]
```

Comment:
Do we need to check the name and version both?

Reply:
Meaning that the package version will have all the dependencies. Not at the name level? So what is the reason to check the name level also?

Reply:
I believe you are correct, at least for now when implementing dependencies only they are attached at the version.  Just changed (as well as removing the nodeType switch you commented on earlier).

Reply:
Yeah, I would have a check based on whether it is required or can be at the "packageName" level. Else you are unnecessarily computing.

Reply:
So in the end I think you would require both the version and the name to be added because some relationships like isDependencies are attached at the packageName level while others like isOccurrences are attached at the package version and both may be relevant for exploring the graph.